src/iterators.jl

```diff
@@ -1,145 +1,190 @@
+    N = nnodes_per_cell(dh.grid)
+    nodes = zeros(Int, N)
+    coords = zeros(Vec{dim,T}, N)
+    n = ndofs_per_cell(dh)
+    celldofs = zeros(Int, n)
+    return CellCache(flags, dh.grid, ScalarWrapper(-1), nodes, coords, dh, celldofs)
 end
 
-CellIterator(grid::Grid{dim,C,T}, cellset::Union{AbstractVector{Int},Nothing}=nothing, flags::UpdateFlags=UpdateFlags()) where {dim,C,T} =
-    CellIterator{dim,C,T}(grid, cellset, flags)
-CellIterator(dh::DofHandler{dim,T}, cellset::Union{AbstractVector{Int},Nothing}=nothing, flags::UpdateFlags=UpdateFlags()) where {dim,T} =
-    CellIterator{dim,getcelltype(dh.grid),T}(dh, cellset, flags)
-CellIterator(dh::MixedDofHandler{dim,T}, cellset::Union{AbstractVector{Int},Nothing}=nothing, flags::UpdateFlags=UpdateFlags()) where {dim,T} =
-    CellIterator{dim,getcelltype(dh.grid),T}(dh, cellset, flags)
-
-# iterator interface
-function Base.iterate(ci::CellIterator, state = 1)
-    if state > (ci.cellset === nothing ? getncells(ci.grid) : length(ci.cellset))
-        return nothing
-    else
-        return (reinit!(ci, state), state+1)
+# TODO: Can always resize and combine the two reinit! methods maybe?
+function reinit!(cc::CellCache, i::Int)
+    cc.cellid[] = i
+    if cc.flags.nodes
+        cellnodes!(cc.nodes, cc.grid, i)
+    end
+    if cc.flags.coords
+        cellcoords!(cc.coords, cc.grid, i)
+    end
+    if cc.dh !== nothing && cc.flags.dofs
+        @assert cc.dh isa DofHandler
+        celldofs!(cc.dofs, cc.dh, i)
     end
+    return cc
 end
-Base.length(ci::CellIterator)  = ci.cellset === nothing ? length(ci.grid.cells) : length(ci.cellset)
-
-Base.IteratorSize(::Type{T})   where {T<:CellIterator} = Base.HasLength() # this is default in Base
-Base.IteratorEltype(::Type{T}) where {T<:CellIterator} = Base.HasEltype() # this is default in Base
-Base.eltype(::Type{T})         where {T<:CellIterator} = T
-
-# utility
-@inline getnodes(ci::CellIterator) = ci.nodes
-@inline getcoordinates(ci::CellIterator) = ci.coords
-@inline nfaces(ci::CellIterator) = nfaces(eltype(ci.grid.cells))
-@inline onboundary(ci::CellIterator, face::Int) = ci.grid.boundary_matrix[face, ci.current_cellid[]]
-@inline cellid(ci::CellIterator) = ci.current_cellid[]
-@inline celldofs!(v::Vector, ci::CellIterator) = celldofs!(v, ci.dh, ci.current_cellid[])
-@inline celldofs(ci::CellIterator) = ci.celldofs
-
-function reinit!(ci::CellIterator{dim,C}, i::Int) where {dim,C}
-    ci.current_cellid[] = ci.cellset === nothing ? i : ci.cellset[i]
-
-    if ci.flags.nodes
-        if ci.dh !== nothing && ci.dh isa MixedDofHandler
-            cellnodes!(ci.nodes, ci.dh, ci.current_cellid[])
-        else
-            cellnodes!(ci.nodes, ci.grid, ci.current_cellid[])
-        end
+function reinit!(cc::CellCache{<:Any,<:AbstractGrid,<:MixedDofHandler}, i::Int)
+    @assert cc.dh isa MixedDofHandler
+    cc.cellid[] = i
+    if cc.flags.nodes
+        resize!(cc.nodes, nnodes_per_cell(cc.dh, i))
```

Comment:
This is new behavior, since previously we would have gotten an error if we loop over a cellset with different cell types. Maybe this is better :thinking: @kimauth ?